File: 222.完全二叉树的节点个数.java
/*
 * @lc app=leetcode.cn id=222 lang=java
 *
 * [222] 完全二叉树的节点个数
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        // time complexity O(logN*logN)
        TreeNode l = root, r = root;
        int hl = 0, hr = 0;
        while (l != null) {
            l = l.left;
            ++hl;
        }
        while (r != null) {
            r = r.right;
            ++hr;
        }
        // At least one of the subtrees of a complete binary tree 
        // is a full binary tree
        return hl == hr ? 
            (int) Math.pow(2, hl) - 1 : 
            1 + countNodes(root.left) + countNodes(root.right);
    }
}
// @lc code=end

File: 222.完全二叉树的节点个数.py
#
# @lc app=leetcode.cn id=222 lang=python3
#
# [222] 完全二叉树的节点个数
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    '''
    # general solution for any binary tree
    def countNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        return self.countNodes(root.left) + self.countNodes(root.right) + 1
    '''

    # Using the properties of complete binary tree
    # T: O(log(n) ^ log(n))
    # S: O(log(n))
    def countNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0
        left = self.getLevels(root.left)
        right = self.getLevels(root.right)
        if left == right:
            # Left subtree is a perfect binary tree
            # Right subtree is a complete binary tree
            return self.countNodes(root.right) + (1 << left)   # 2 ^ left
        
        # left > right
        # Left subtree is a complete binary tree (deeper)
        # Right subtree is a perfect binary tree
        return self.countNodes(root.left) + (1 << right)  # 2 ^ right

    # nodes in the last level are as far left as possible
    @staticmethod
    def getLevels(root):
        lvl = 0
        while root is not None:
            lvl += 1
            root = root.left
        return lvl

# @lc code=end

